File: packages/sdk/src/core/utils/logger.ts
import type { LogLevel } from '../types';

const LOG_LEVELS: Record<LogLevel, number> = {
  info: 0,
  warn: 1,
  error: 2,
  none: 3,
};

class Logger {
  private level = LOG_LEVELS.none;

  setLogLevel(level: LogLevel): void {
    this.level = LOG_LEVELS[level];
  }

  info(message: string, metadata?: Record<string, unknown>): void {
    if (this.level <= LOG_LEVELS.info) {
      if (metadata && Object.keys(metadata).length > 0) {
        console.log(`[Phase] ${message}`, metadata);
      } else {
        console.log(`[Phase] ${message}`);
      }
    }
  }

  warn(message: string, metadata?: Record<string, unknown>): void {
    if (this.level <= LOG_LEVELS.warn) {
      if (metadata && Object.keys(metadata).length > 0) {
        console.warn(`[Phase] ${message}`, metadata);
      } else {
        console.warn(`[Phase] ${message}`);
      }
    }
  }

  error(message: string, error?: unknown): void {
    if (this.level <= LOG_LEVELS.error) {
      if (error instanceof Error) {
        console.error(`[Phase] ${message}:`, error.message);
      } else if (error) {
        console.error(`[Phase] ${message}:`, error);
      } else {
        console.error(`[Phase] ${message}`);
      }
    }
  }
}

export const logger = new Logger();

File: packages/swift/Sources/PhaseAnalytics/Core/Utils/Logger.swift
import Foundation

internal final class Logger: Sendable {
    private let levelValue: ThreadSafeLock<Int>

    private let levels: [LogLevel: Int] = [
        .info: 0,
        .warn: 1,
        .error: 2,
        .none: 3,
    ]

    init() {
        self.levelValue = ThreadSafeLock(3)
    }

    func setLogLevel(_ level: LogLevel) {
        levelValue.withLock { $0 = levels[level] ?? 3 }
    }

    func info(_ message: String) {
        levelValue.withLock { currentLevel in
            guard currentLevel <= (levels[.info] ?? 0) else { return }
            print("[Phase] \(message)")
        }
    }

    func warn(_ message: String) {
        levelValue.withLock { currentLevel in
            guard currentLevel <= (levels[.warn] ?? 1) else { return }
            print("[Phase] \(message)")
        }
    }

    func error(_ message: String, _ error: Error? = nil) {
        levelValue.withLock { currentLevel in
            guard currentLevel <= (levels[.error] ?? 2) else { return }
            if let error = error {
                print("[Phase] \(message): \(error.localizedDescription)")
            } else {
                print("[Phase] \(message)")
            }
        }
    }
}

internal let logger = Logger()
